test/e2e/annotations/canary.go

```diff
@@ -105,12 +113,19 @@ var _ = framework.DescribeAnnotation("canary-*", func() {
 				80,
 				canaryAnnotations))
 
-			f.HTTPTestClient().
-				GET("/info").
-				WithHeader("Host", host).
-				WithHeader("CanaryByHeader", "always").
-				Expect().
-				Status(http.StatusNotFound)
+			f.WaitForNginxServer(host,
+				func(server string) bool {
+					return strings.Contains(server, fmt.Sprintf("server_name %v", host))
+				})
+
+			for _, wh := range wrongHeaders {
+				f.HTTPTestClient().
+					GET("/info").
+					WithHeader("Host", host).
+					WithHeader(wh, "always").
+					Expect().
+					Status(http.StatusNotFound)
+			}
 		})
```

Comment:
what here was fixed:

- added `f.WaitForNginxServer...` for fixing the race condition between `f.EnsureIngress` and `f.HTTPTestClient()...Expect().Status(http.StatusNotFound)`
- test `should return 404 status for requests to the canary if no matching ingress is found` has been fixed (a different header is being sent than the one that is configured)
- added several options for checking wrong headers